src/validation/func.rs

```diff
@@ -1598,14 +1372,19 @@ fn make_top_frame_polymorphic(
 	let frame = frame_stack
 		.top_mut()
 		.expect("make_top_frame_polymorphic is called with empty frame stack");
-	value_stack.resize(frame.value_stack_len, StackValueType::Any);
+
+	// shrink value stack to match top frame
+	debug_assert!(frame.value_stack_len <= value_stack.len());
+	while value_stack.len() > frame.value_stack_len {
+		value_stack.pop().expect(
+			"frame.value_stack_len >= 0, value_stack.len() > frame.value_stack_len :. value_stack.len() > 0; qed",
```

Comment:
I know we've got terrible "... qed" error messages everywhere, but can you make this from the perspective of if a user ever actually hit it? So something like, "The value stack is empty, this should never happen since `frame.value_stack_len >= 0` and `value_stack.len() > frame.value_stack_len`, so therefore `value_stack.len()` must be more than 0". "Impossible" cases do happen, we've had users come to us with "qed" error messages that they're really confused by since this style of message actually tells you the _opposite_ of what is true (since they're designed to describe the case in which the `expect` did not fire).

Reply:
done